syne_tune/optimizer/schedulers/searchers/kde/kde_searcher.py

```diff
@@ -353,30 +354,27 @@ def acquisition_function(x):
 
         return suggestion
 
-    def _train_kde(self, train_data, train_targets):
-        if train_data.shape[0] < self.num_min_data_points:
+    def _good_data_size(self, data_shape: Tuple[int, int]) -> Optional[int]:
+        num_data, num_features = data_shape
+        n_good = max(self.num_min_data_points, (self.top_n_percent * num_data) // 100)
+        if min(n_good, num_data - n_good) <= num_features:
+            return None
+        else:
+            return n_good
+
+    def _train_kde(
+        self, train_data: np.ndarray, train_targets: np.ndarray
+    ) -> Optional[Tuple[Any, Any]]:
+        train_data = train_data.reshape((train_targets.size, -1))
+        n_good = self._good_data_size(train_data.shape)
+        if n_good is None:
             return None
-
-        n_good = max(
-            self.num_min_data_points, (self.top_n_percent * train_data.shape[0]) // 100
-        )
-        n_bad = max(
-            self.num_min_data_points,
-            ((100 - self.top_n_percent) * train_data.shape[0]) // 100,
-        )
 
         idx = np.argsort(train_targets)
-
         train_data_good = train_data[idx[:n_good]]
-        train_data_bad = train_data[idx[n_good : n_good + n_bad]]
-
-        if train_data_good.shape[0] <= train_data_good.shape[1]:
-            return None
-        if train_data_bad.shape[0] <= train_data_bad.shape[1]:
-            return None
+        train_data_bad = train_data[idx[n_good:]]
```

Comment:
Why do we not check anymore if n_bad is greater than self.num_min_data_points?

Reply:
This is done in the `_good_data_size` method